state-sync/src/coordinator.rs

```diff
@@ -1146,69 +1227,64 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
-        // notify consensus if sync request timed out
-        if sync_request_expired {
-            counters::SYNC_REQUEST_RESULT
-                .with_label_values(&[counters::TIMEOUT_LABEL])
-                .inc();
-            warn!(LogSchema::event_log(
-                LogEntry::SyncRequest,
-                LogEvent::Timeout
-            ));
+        // Check if the sync request has timed out (i.e., if we aren't committing fast enough)
+        if let Some(sync_request) = self.sync_request.as_ref() {
+            let timeout_between_commits =
+                Duration::from_millis(self.config.sync_request_timeout_ms);
+            let commit_deadline = sync_request
+                .last_commit_timestamp
+                .checked_add(timeout_between_commits)
+                .ok_or_else(|| {
+                    Error::IntegerOverflow("The commit deadline timestamp has overflown!".into())
+                })?;
+            if SystemTime::now().duration_since(commit_deadline).is_ok() {
+                // Timeout exceeded!
+                counters::SYNC_REQUEST_RESULT
+                    .with_label_values(&[counters::TIMEOUT_LABEL])
+                    .inc();
+                warn!(LogSchema::event_log(
+                    LogEntry::SyncRequest,
+                    LogEvent::Timeout
+                ));
 
-            if let Some(sync_request) = self.sync_request.take() {
-                if let Err(e) = Self::send_sync_req_callback(
-                    sync_request,
-                    Err(format_err!("request timed out")),
-                ) {
-                    error!(
-                        LogSchema::event_log(LogEntry::SyncRequest, LogEvent::CallbackFail)
-                            .error(&e)
-                    );
+                if let Some(sync_request) = self.sync_request.take() {
+                    if let Err(e) = Self::send_sync_req_callback(
+                        sync_request,
+                        Err(format_err!("Sync request timed out!")), // TODO(joshlind): fix these callback return messages!
+                    ) {
+                        error!(
+                            LogSchema::event_log(LogEntry::SyncRequest, LogEvent::CallbackFail)
+                                .error(&e)
+                        );
+                    }
                 }
             }
         }
 
+        // If the coordinator didn't make progress by the expected time or did not
+        // send a request for the current local synced version, issue a new request.
         let known_version = self.local_state.synced_version();
-
-        // if coordinator didn't make progress by expected time or did not send a request for current
-        // local synced version, issue new request
-        if self
-            .request_manager
-            .check_timeout(known_version)
-            .unwrap_or(false)
-        {
-            // log and count timeout
+        let trusted_epoch = self.local_state.trusted_epoch();
```

Comment:
Separate variable because?

Also, now it's making a function call before the if branch, but it's not used outside of it.

Reply:
- Separate variable because the `if let Err(e)` expression is really long, so breaking it down for clarity.
- Moving this out of the loop isn't absolutely required, but it's code grouping (i.e., get all of our local state first, before we do processing).
- Regarding the performance implications of this, it's non-existent (this is a premature optimization). The compiler will detect this (it is very aggressive).

Reply:
Heh, I just don't trust compilers all the time, sometimes you can easily make them do different things with slight changes.

Let's compromise, can you move the variable within the outer if statement?

Reply:
Sure, I can do that 😄 Thanks!